### DISCUSSION TITLE
Question on reading DateTime strings

### DISCUSSION THREAD

**Discussion by champlin2** [JoshClose/CsvHelper#1748] [Q&A]
When reading a field that is specified as a DateTime if the CSV only contains a date the time portion is set to midnight (00:00:00).  Is there any way to tell if the CSV just contained the date portion (and the time was set to midnight automatically on reading) or if the CSV actually contained a time portion of 00:00:00?

**Comment by JoshClose** [comment]
Where would you be looking at the data? You could create an extra property like `ContainsTime` and set that based on the value of that field. You'd do that either through a type converter, or in a mapping using the `Convert` map method.

**Reply by champlin2** [reply]
@JoshClose 

Ok tried doing this by creating a custom converter to determine if the date had a time portion or if it was just a date and the time was assumed to be midnight.  That is all working.  Then I added a public bool member and a public method to set the value or the member to true and then passed this as an Action to the custom converter constructor.  It calls back the Action when it detects that the time portion was not set correctly and sets the public member but later on when I access the public member of the class its false - like it gets lost after the converter sets it.  Any ideas?

```cs
public sealed class TestClass : ClassMap<TestClass>
{
  public TestClass()
  {
    ...
    Map(m => m.EndDate).Name("EndDate").TypeConverter(new CustomEndDateConverter(EndDateWasIncremented));
    ...
  }

  ...
  public DateTime? EndDate {get; set;}
  ...
  public bool EndDateIncremented {get; set;}
  ...
  public void EndDateWasIncremented() { EndDateIncremented = true; }
}

public class CustomEndDateConverter : DateTimeConverter
{
  private readonly Action _endDateIncrementedAction;

  public CustomEndDateConverter(Action endDateIncrementedAction)
  {
    _endDateIncrementedAction = endDateIncrementedAction;
  }

  public override object ConvertFromString(string text, IReaderRow row, MemberMapData memberMapData)
  {
    // If no time portion entered
    _endDateIncrementedAction?.Invoke();
  }
}
```

**Comment by JoshClose** [comment]
I'm not sure how that compiles as `ConvertFromString` needs to have an object returned from it.

Here is an example.

```cs
void Main()
{
    var s = new StringBuilder();
    s.Append($"Id,Name,EndDate\r\n");
    s.Append($"1,one,{new DateTime(2021, 4, 13)}\r\n");
    var config = new CsvConfiguration(CultureInfo.InvariantCulture)
    {
    };
    using (var reader = new StringReader(s.ToString()))
    using (var csv = new CsvReader(reader, config))
    {
        csv.Context.RegisterClassMap<FooMap>();
        csv.GetRecords<Foo>().ToList().Dump();
    }
}

private class Foo
{
    public int Id { get; set; }
    public string Name { get; set; }
    public DateTime? EndDate { get; set; }
    public bool IsEndDateMidnight { get; set; }
}

private class FooMap : ClassMap<Foo>
{
    public FooMap()
    {
        Map(m => m.Id);
        Map(m => m.Name);
        Map(m => m.EndDate);
        Map(m => m.IsEndDateMidnight).Convert(args =>
        {
           var endDate = args.Row.GetField<DateTime?>(nameof(Foo.EndDate));
           if (endDate == null)
           {
               return false;
           }
           
           return endDate.Value.Hour == 0 && endDate.Value.Minute == 0 && endDate.Value.Second == 0 && endDate.Value.Millisecond == 0;
        });
    }
}
```

**Reply by champlin2** [reply]
@JoshClose 

I was just stripping down the code to the relevant parts (posted the full version below).  I think in your example above the IsEndDateMidnight is calculated after EndDate has been read from the CSV file.  The problem there is we don't know if the CSV contained "01/01/2021" (and midnight was assumed) or "01/01/2021 00:00:00" (and midnight was explicitly specified).  That's what I'm trying to distinguish.  Here's the full code of what I described above that isn't working:

````
public sealed class TestClass : ClassMap<TestClass>
    {
        private bool _endDateIncremented;

        public TestClass()
        {
            // EndDate has a custom converter that determines if a time component was entered which will be used if it was
            // otherwise will use midnight of the next day to get all messages with a timestamp of the day entered

            Map(m => m.EndDate).Name("EndDate").TypeConverter(new CustomEndDateConverter(EndDateIncremented));
        }

        public DateTime? EndDate { get; set; }

        public void EndDateIncremented()
        {
            _endDateIncremented = true;
        }

        public bool EndDateWasIncremented()
        {
            return _endDateIncremented;
        }
    }

public class CustomEndDateConverter : DateTimeConverter
    {
        private readonly Action _endDateIncrementedAction;

        public CustomEndDateConverter(Action endDateIncrementedAction)
        {
            _endDateIncrementedAction = endDateIncrementedAction;
        }

        public override object ConvertFromString(string text, IReaderRow row, MemberMapData memberMapData)
        {
            if (string.IsNullOrEmpty(text)) return null;

            // Attempt to parse the date with the various formats and if successful see if any of the time components are
            // part of the format that was used for parsing.

            // Adapted from https://stackoverflow.com/questions/17204289/how-to-know-if-time-part-was-present-when-parsing-with-datetime

            // These character format codes come from here:
            // http://msdn.microsoft.com/en-us/library/8kb3ddd4.aspx

            var dateTimePatterns = CultureInfo.CurrentUICulture.DateTimeFormat.GetAllDateTimePatterns();
            var timeComponentPresent = (from pattern in dateTimePatterns
                                        where DateTime.TryParseExact(text,
                                            pattern,
                                            CultureInfo.InvariantCulture,
                                            DateTimeStyles.AllowWhiteSpaces | DateTimeStyles.AssumeLocal,
                                            out _)
                                        select pattern.IndexOfAny(new[] { 'H', 'h', 'm', 's', 'f', 'F', 't', 'z' }) != -1)
                .FirstOrDefault();

            var readDateTime = base.ConvertFromString(text, row, memberMapData);

            if (timeComponentPresent)
                return readDateTime;

            // If no date was entered make the value midnight of the next day
            readDateTime = ((DateTime?)readDateTime)?.Date.AddDays(1);
            _endDateIncrementedAction?.Invoke();

            return readDateTime;
        }
    }
````

**Reply by champlin2** [reply]
@JoshClose just wondering if you saw the same behavior after calling GetRecords with my sample code above

**Reply by champlin2** [reply]
@JoshClose 

Tried using the ".Convert" option instead of TypeConverter (and add a method to the custom converter to see if we incremented the date) but not sure what to pass to the ConvertFromString method:


            Map(m => m.EndDate).Name("EndDate").Convert(args =>
            {
                var endDateConverter = new CustomEndDateConverter();
                var endDate = endDateConverter.ConvertFromString( ??, ??, ??);
                _endDateIncremented = endDateConverter.EndDateWasIncremented();
                return (DateTime?)endDate;
            });

**Reply by JoshClose** [reply]
If you want to use a customer converter, just do
```cs
Map(m => m.EndDate).Name("EndDate").TypeConverter<CustomEndDateConverter>();
```

**Reply by champlin2** [reply]
@JoshClose 

I tried that (see above) but as I described when I called the callback to set the property in the TestClass it doesn't stick after calling GetRecords<>

**Reply by JoshClose** [reply]
I'm confused. You're setting a value on the mapping?

**Reply by champlin2** [reply]
@JoshClose what I'm trying to do is when the value is converted from a string in the CSV file to be able to tell if the CSV contained 00:00:00 for the time portion of the DateTime or if no time portion was entered and 00:00;00 was assumed.  Later on when I'm iterating over the list of objects that is returned from GetRecords I want to know which had midnight set and which it was assumed for so I'm trying to set a member in that TestClass to let me know which of the two it was.
